opennebula/helpers_vm.go

```diff
@@ -137,45 +138,93 @@ func vmDiskResize(vmc *goca.VMController, timeout, diskID, newsize int) error {
 }
 
 // vmNICAttach is an helper that synchronously attach a nic
-func vmNICAttach(vmc *goca.VMController, timeout int, nicTpl *shared.NIC) error {
+func vmNICAttach(vmc *goca.VMController, timeout int, nicTpl *shared.NIC) (int, error) {
 
 	networkID, err := nicTpl.GetI(shared.NetworkID)
 	if err != nil {
-		return fmt.Errorf("NIC template doesn't have a network ID")
+		return -1, fmt.Errorf("NIC template doesn't have a network ID")
 	}
 
 	log.Printf("[DEBUG] Attach NIC to network (ID:%d)", networkID)
 
+	// Retrieve NIC list
+	vm, err := vmc.Info(false)
+	if err != nil {
+		return -1, err
+	}
+
+	set := schema.NewSet(schema.HashString, []interface{}{})
+	for _, nic := range vm.Template.GetNICs() {
+		set.Add(nic.String())
+	}
+
 	err = vmc.AttachNIC(nicTpl.String())
 	if err != nil {
-		return fmt.Errorf("can't attach network with ID:%d: %s\n", networkID, err)
+		return -1, fmt.Errorf("can't attach network with ID:%d: %s\n", networkID, err)
 	}
 
 	// wait before checking NIC
 	_, err = waitForVMState(vmc, timeout, vmNICUpdateReadyStates...)
 	if err != nil {
-		return fmt.Errorf(
+		return -1, fmt.Errorf(
 			"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmNICUpdateReadyStates, " "), err)
 	}
 
-	// Check that NIC is attached
-	vm, err := vmc.Info(false)
+	// compare NIC list to check that a new NIC is attached
+	vm, err = vmc.Info(false)
 	if err != nil {
-		return err
+		return -1, err
 	}
 
-	for _, attachedNic := range vm.Template.GetNICs() {
+	oldNICs := make([]shared.NIC, 0, 1)
+	for _, nic := range vm.Template.GetNICs() {
 
-		attachedNicNetworkID, _ := attachedNic.GetI(shared.NetworkID)
-		if attachedNicNetworkID == networkID {
-			return nil
+		if set.Contains(nic.String()) {
+			continue
 		}
+
+		oldNICs = append(oldNICs, nic)
 	}
 
-	// If NIC not attached, retrieve error message
-	vmerr, _ := vm.UserTemplate.Get(vmk.Error)
+	var attachedNIC *shared.NIC
+
+	switch len(oldNICs) {
```

Comment:
I don't understand the usage of switch here. 
Why not testing len to 0 and in other case doing the loop (event for len=1) ?

Reply:
Among others, It seems that I used some misleading names in this function.
`set` keeps NICs presents before the update and I probably should rename it `refNICs` 
`oldNICs` contains the NIC that are not in `refNICSet`. I should rename it `updatedNICs`

Indeed, I'll try to simplify this function.

Reply:
It seems I also used misleading names for `set` and `oldNICs`.

Indeed, I'll update this for something simpler and more reliable